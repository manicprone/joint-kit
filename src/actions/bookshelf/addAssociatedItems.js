import * as StatusErrors from '../../core/errors/status-errors'
import INSTANCE from '../../core/constants/instance-constants'
import ACTION from '../../core/constants/action-constants'
import getItem from './getItem'
import getItems from './getItems'
import { handleDataResponse, handleErrorResponse } from './handlers/response-handlers'

const debug = false

export default async function addAssociatedItems(joint, spec = {}, input = {}, output) {
  const bookshelf = joint[INSTANCE.PROP_SERVICE]
  const trx = input[ACTION.INPUT_TRANSACTING]

  // Continue on existing transaction...
  if (trx) return performAddAssociatedItems(joint, spec, input, output)

  // Otherwise, start new transaction...
  return bookshelf.transaction((newTrx) => {
    const newInput = Object.assign({}, input)
    newInput[ACTION.INPUT_TRANSACTING] = newTrx
    return performAddAssociatedItems(joint, spec, newInput, output)
  })
}

async function performAddAssociatedItems(joint, spec = {}, input = {}, output) {
  const specMain = spec[ACTION.RESOURCE_MAIN]
  const modelNameMain = (specMain) ? specMain[ACTION.SPEC_MODEL_NAME] : null
  const specAssoc = spec[ACTION.RESOURCE_ASSOCIATION]
  const assocName = (specAssoc) ? specAssoc[ACTION.SPEC_ASSOCIATION_NAME] : null
  const inputMain = input[ACTION.RESOURCE_MAIN]
  const inputAssoc = input[ACTION.RESOURCE_ASSOCIATION]
  const trx = input[ACTION.INPUT_TRANSACTING]

  // Reject when required properties are not provided...
  const missingProps = []
  if (!specMain) missingProps.push(`spec.${ACTION.RESOURCE_MAIN}`)
  if (!specAssoc) missingProps.push(`spec.${ACTION.RESOURCE_ASSOCIATION}`)
  if (!assocName) missingProps.push(`spec.${ACTION.RESOURCE_ASSOCIATION}.${ACTION.SPEC_ASSOCIATION_NAME}`)
  if (!inputMain) missingProps.push(`input.${ACTION.RESOURCE_MAIN}`)
  if (!inputAssoc) missingProps.push(`input.${ACTION.RESOURCE_ASSOCIATION}`)
  if (missingProps.length > 0) {
    if (debug) console.log(`[JOINT] [action:addAssociatedItems] Required properties missing: "${missingProps.join('", "')}"`)
    return Promise.reject(StatusErrors.generateInvalidAssociationPropertiesError(missingProps))
  }

  // Lookup model name of association, add to spec if not provided...
  let modelNameAssoc = specAssoc[ACTION.SPEC_MODEL_NAME]
  if (!modelNameAssoc) {
    modelNameAssoc = (joint.modelNameOfAssoc[modelNameMain]) ? joint.modelNameOfAssoc[modelNameMain][assocName] : null
    specAssoc[ACTION.SPEC_MODEL_NAME] = modelNameAssoc
  }

  // Load trx to both resources...
  inputMain[ACTION.INPUT_TRANSACTING] = trx
  inputAssoc[ACTION.INPUT_TRANSACTING] = trx

  // Include existing associations of this type (for return)...
  inputMain[ACTION.INPUT_ASSOCIATIONS] = [assocName]

  try {
    // Lookup resources...
    const main = await getItem(joint, specMain, inputMain)
    const assoc = await getItems(joint, specAssoc, inputAssoc)

    // TODO: It looks like the latest version of Bookshelf no longer guards against duplicate assoc (via attach) !!!
    //       We will have to provide the logic here !!!
    // const existingAssoc = main.relations[assocName]
    // console.log('[DEV] existing assocs =>', existingAssoc.models)
    // console.log('[DEV] assocs to add =>', assoc.models)

    // Reject with 404 if instances of the requested association were not found...
    if (assoc.length === 0) {
      return Promise.reject(StatusErrors.generateAssociatedItemsDoNotExistError(modelNameAssoc))
    }

    // Otherwise, attach associations to main...
    await main.related(assocName).attach(assoc.models, { transacting: trx })
    return handleDataResponse(joint, modelNameMain, main, output)

  } catch (error) {
    return handleErrorResponse(error, 'addAssociatedItems', modelNameMain)
  }
} // END - performAddAssociatedItems
