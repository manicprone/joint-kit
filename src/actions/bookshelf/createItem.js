import objectUtils from '../../utils/object-utils'
import * as StatusErrors from '../../core/errors/status-errors'
import * as AuthUtils from '../../core/authorization/auth-utils'
import INSTANCE from '../../core/constants/instance-constants'
import ACTION from '../../core/constants/action-constants'
import * as ActionUtils from '../action-utils'
import { handleDataResponse, handleErrorResponse } from './handlers/response-handlers'

const debug = false

export default async function createItem (joint, spec = {}, input = {}, output) {
  const bookshelf = joint[INSTANCE.PROP_SERVICE]
  const trx = input[ACTION.INPUT_TRANSACTING]

  // Continue on existing transaction...
  if (trx) return performCreateItem(joint, spec, input, output)

  // Otherwise, start new transaction...
  return bookshelf.transaction((newTrx) => {
    const newInput = Object.assign({}, input)
    newInput[ACTION.INPUT_TRANSACTING] = newTrx
    return performCreateItem(joint, spec, newInput, output)
  })
}

async function performCreateItem (joint, spec = {}, input = {}, output) {
  const bookshelf = joint[INSTANCE.PROP_SERVICE]
  const modelName = spec[ACTION.SPEC_MODEL_NAME]
  const specFields = spec[ACTION.SPEC_FIELDS]
  const specAuth = spec[ACTION.SPEC_AUTH] || {}
  const authRules = specAuth[ACTION.SPEC_AUTH_RULES]
  const inputFields = ActionUtils.prepareFieldData(specFields, input[ACTION.INPUT_FIELDS])
  const trx = input[ACTION.INPUT_TRANSACTING]
  const authContext = input[ACTION.INPUT_AUTH_CONTEXT]

  // Reject if model does not exist...
  const model = bookshelf.model(modelName)
  if (!model) {
    if (debug) console.log(`[JOINT] [action:createItem] The model "${modelName}" is not recognized`)
    return Promise.reject(StatusErrors.generateModelNotRecognizedError(modelName))
  }

  // Reject when required fields are not provided...
  const requiredFieldCheck = ActionUtils.checkRequiredFields(specFields, inputFields)
  if (!requiredFieldCheck.satisfied) {
    if (debug) console.log('[JOINT] [action:createItem] Action has missing required fields:', requiredFieldCheck.missing)
    return Promise.reject(StatusErrors.generateMissingFieldsError(requiredFieldCheck.missing))
  }

  // Respect auth...
  if (authRules) {
    const ownerCreds = ActionUtils.parseOwnerCreds(specAuth, inputFields)
    if (!AuthUtils.isAllowed(authContext, authRules, ownerCreds)) {
      return Promise.reject(StatusErrors.generateNotAuthorizedError())
    }
  } // end-if (authRules)

  // Prepare action options...
  const actionOpts = {}
  if (trx) actionOpts.transacting = trx

  // Build create package...
  const createData = {}
  if (inputFields && specFields) {
    specFields.forEach((fieldSpec) => {
      const fieldName = fieldSpec.name
      const hasDefault = objectUtils.has(fieldSpec, ACTION.SPEC_FIELDS_OPT_DEFAULT_VALUE)
      const defaultValue = (hasDefault) ? ActionUtils.processDefaultValue(inputFields, fieldSpec[ACTION.SPEC_FIELDS_OPT_DEFAULT_VALUE]) : null
      const hasInput = objectUtils.has(inputFields, fieldName)
      const isLocked = objectUtils.get(fieldSpec, ACTION.SPEC_FIELDS_OPT_LOCKED, false)

      if (hasInput || hasDefault) {
        const inputValue = !isLocked && hasInput ? inputFields[fieldName].value : defaultValue
        createData[fieldName] = inputValue
      }
    }) // end-specFields.forEach
  } // end-if (inputFields && specFields)

  // Debug executing logic...
  if (debug) console.log(`[JOINT] [action:createItem] EXECUTING => CREATE ${modelName} WITH`, createData)

  try {
    // Create item...
    const data = await model.forge(createData).save(null, actionOpts)
    return handleDataResponse(joint, modelName, data, output)
  } catch (error) {
    return handleErrorResponse(error, 'createItem', modelName)
  }
} // END - performCreateItem
