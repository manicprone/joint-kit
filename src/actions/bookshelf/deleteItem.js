import objectUtils from '../../utils/object-utils'
import * as StatusErrors from '../../core/errors/status-errors'
import * as AuthUtils from '../../core/authorization/auth-utils'
import INSTANCE from '../../core/constants/instance-constants'
import ACTION from '../../core/constants/action-constants'
import * as ActionUtils from '../action-utils'
import * as BookshelfUtils from './utils/bookshelf-utils'
import { handleDataResponse, handleErrorResponse } from './handlers/response-handlers'

const debug = false

export default async function deleteItem (joint, spec = {}, input = {}, output) {
  const bookshelf = joint[INSTANCE.PROP_SERVICE]
  const trx = input[ACTION.INPUT_TRANSACTING]

  // Continue on existing transaction...
  if (trx) return performDeleteItem(joint, spec, input, output)

  // Otherwise, start new transaction...
  return bookshelf.transaction((newTrx) => {
    const newInput = Object.assign({}, input)
    newInput[ACTION.INPUT_TRANSACTING] = newTrx
    return performDeleteItem(joint, spec, newInput, output)
  })
}

async function performDeleteItem (joint, spec = {}, input = {}, output) {
  const bookshelf = joint[INSTANCE.PROP_SERVICE]
  const modelName = spec[ACTION.SPEC_MODEL_NAME]
  const specFields = ActionUtils.normalizeFieldSpec(spec[ACTION.SPEC_FIELDS])
  const specAuth = spec[ACTION.SPEC_AUTH] || {}
  const inputFields = ActionUtils.prepareFieldData(specFields, input[ACTION.INPUT_FIELDS])
  const trx = input[ACTION.INPUT_TRANSACTING]
  const authContext = input[ACTION.INPUT_AUTH_CONTEXT]

  // Reject if model does not exist...
  const model = bookshelf.model(modelName)
  if (!model) {
    if (debug) console.log(`[JOINT] [action:deleteItem] The model "${modelName}" is not recognized`)
    return Promise.reject(StatusErrors.generateModelNotRecognizedError(modelName))
  }

  // Reject when required fields are not provided...
  const requiredFieldCheck = ActionUtils.checkRequiredFields(specFields, inputFields)
  if (!requiredFieldCheck.satisfied) {
    if (debug) console.log('[JOINT] [action:deleteItem] Action has missing required fields:', requiredFieldCheck.missing)
    return Promise.reject(StatusErrors.generateMissingFieldsError(requiredFieldCheck.missing))
  }

  // Perform lookup of item first, if specified...
  const lookupFieldData = ActionUtils.getLookupFieldData(specFields, inputFields)
  if (lookupFieldData) {
    return doLookupThenAction(joint, lookupFieldData, modelName, specFields, specAuth, inputFields, authContext, trx, output)
  }

  // Otherwise, just perform action...
  return doAction(joint, modelName, specFields, specAuth, null, inputFields, authContext, trx, output)
} // END - performDeleteItem

async function doLookupThenAction (joint, lookupFieldData, modelName, specFields, specAuth, inputFields, authContext, trx, output) {
  const bookshelf = joint[INSTANCE.PROP_SERVICE]
  const getItemOpts = { require: true }
  if (trx) getItemOpts.transacting = trx

  try {
    const resource = await bookshelf.model(modelName).query((queryBuilder) => {
      Object.entries(lookupFieldData)
        .forEach(([fieldName, field]) => {
          BookshelfUtils.appendWhereClause(queryBuilder, fieldName, field.value, field.matchStrategy)
        })
    }).fetch(getItemOpts)

    // Prepare ownerCreds from retrieved data...
    const combinedFields = Object.assign({}, resource.attributes, inputFields)
    const ownerCreds = ActionUtils.parseOwnerCreds(specAuth, combinedFields)

    return doAction(joint, modelName, specFields, specAuth, ownerCreds, inputFields, authContext, trx, output)
  } catch (error) {
    return handleErrorResponse(error, 'deleteItem', modelName)
  }
} // END - doLookupThenAction

async function doAction (joint, modelName, specFields, specAuth, ownerCreds, inputFields, authContext, trx, output) {
  const bookshelf = joint[INSTANCE.PROP_SERVICE]

  // Respect auth...
  const authRules = specAuth[ACTION.SPEC_AUTH_RULES]
  if (authRules) {
    const ownerCredsData = ownerCreds || ActionUtils.parseOwnerCreds(specAuth, inputFields)
    if (!AuthUtils.isAllowed(authContext, authRules, ownerCredsData)) {
      return Promise.reject(StatusErrors.generateNotAuthorizedError())
    }
  } // end-if (authRules)

  // Prepare action options...
  const actionOpts = { require: true }
  if (trx) actionOpts.transacting = trx

  // Build where clause...
  const queryOpts = (queryBuilder) => {
    if (inputFields && specFields) {
      specFields.forEach((fieldSpec) => {
        const fieldName = fieldSpec.name
        const hasInput = objectUtils.has(inputFields, fieldName)
        const matchStrategy = objectUtils.get(inputFields,
          `${fieldName}.matchStrategy`,
          ACTION.INPUT_FIELD_MATCHING_STRATEGY_EXACT
        )
        if (hasInput) {
          BookshelfUtils.appendWhereClause(queryBuilder, fieldName, inputFields[fieldName].value, matchStrategy)
        }
      })
    } // end-if (inputFields && specFields)
  }

  // Debug executing logic...
  if (debug) console.log(`[JOINT] [action:deleteItem] EXECUTING => DELETE ${modelName} WHERE`, queryOpts)

  try {
    // Delete item...
    const data = await bookshelf.model(modelName).query(queryOpts).destroy(actionOpts)
    return handleDataResponse(joint, modelName, data, output)
  } catch (error) {
    return handleErrorResponse(error, 'deleteItem', modelName)
  }
} // END - doAction
