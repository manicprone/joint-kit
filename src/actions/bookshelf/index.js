// ---------------
// Base actions...
// ---------------
export { default as createItem } from './createItem'
export { default as upsertItem } from './upsertItem'
export { default as updateItem } from './updateItem'
export { default as updateMany } from './updateMany'
export { default as getItem } from './getItem'
export { default as getItems } from './getItems'
export { default as deleteItem } from './deleteItem'

// ----------------------
// Association actions...
// ----------------------
export { default as addAssociatedItems } from './addAssociatedItems'
export { default as hasAssociatedItem } from './hasAssociatedItem'
export { default as getAllAssociatedItems } from './getAllAssociatedItems'
export { default as removeAssociatedItems } from './removeAssociatedItems'
export { default as removeAllAssociatedItems } from './removeAllAssociatedItems'
