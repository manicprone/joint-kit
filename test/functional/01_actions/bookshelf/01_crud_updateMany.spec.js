import { afterAll, beforeAll, beforeEach, describe, expect, it, vi } from 'vitest'
import { omit } from 'lodash/fp'
import ACTION from '../../../../src/core/constants/action-constants'
import Joint from '../../../../src'
import appMgmtModels from '../../../scenarios/app-mgmt/model-config'
import projectAppModels from '../../../scenarios/project-app/model-config'
import blogAppModels from '../../../scenarios/blog-app/model-config'
import bookshelf from '../../../db/bookshelf/service'
import { resetDB } from '../../../db/bookshelf/db-utils'

// remove bookshelf internal fields
const omitInternalFields = omit(['attributes', '_previousAttributes', 'changed'])

let appMgmt = null
let appMgmtJsonApi = null
let projectApp = null
let projectAppJsonApi = null
let blogApp = null
let blogAppJsonApi = null

// -----------------------------------------------------------------------------
// BOOKSHELF ACTIONS (CRUD)
// -----------------------------------------------------------------------------
describe('CRUD ACTIONS [bookshelf]', () => {
  beforeAll(() => {
    vi.useFakeTimers()
    vi.setSystemTime(new Date('2024-01-01T00:00:00.000Z'))

    // --------
    // App Mgmt
    // --------
    appMgmt = new Joint({ service: bookshelf })
    appMgmt.generate({ modelConfig: appMgmtModels, log: false })

    appMgmtJsonApi = new Joint({ service: bookshelf, output: 'json-api' })
    appMgmtJsonApi.generate({ modelConfig: appMgmtModels, log: false })

    // -----------
    // Project App
    // -----------
    projectApp = new Joint({ service: bookshelf })
    projectApp.generate({ modelConfig: projectAppModels, log: false })

    projectAppJsonApi = new Joint({ service: bookshelf, output: 'json-api' })
    projectAppJsonApi.generate({ modelConfig: projectAppModels, log: false })

    // --------
    // Blog App
    // --------
    blogApp = new Joint({
      service: bookshelf,
      settings: {
        auth: {
          debugBuild: false,
          debugCheck: false
        }
      }
    })
    blogApp.generate({ modelConfig: blogAppModels, log: false })

    blogAppJsonApi = new Joint({ service: bookshelf, output: 'json-api' })
    blogAppJsonApi.generate({ modelConfig: blogAppModels, log: false })
  })

  beforeEach(async () => {
    vi.setSystemTime(new Date('2024-01-01T00:00:00.000Z'))
    await resetDB(['profiles', 'projects'])
  })

  afterAll(() => { vi.useRealTimers() })

  describe('updateMany', () => {
    it('should return an error (400) when the input does not provide a "lookup" field', async () => {
      const spec = {
        modelName: 'Project',
        fields: [
          { name: 'id', type: 'Number', required: true, lookup: true },
          { name: 'name', type: 'String' },
          { name: 'brief_description', type: 'String' }
        ]
      }
      const input = {
        fields: {
          name: 'Updated Name'
        }
      }

      await expect(projectApp.updateMany(spec, input))
        .rejects
        .toMatchInlineSnapshot(`
         {
           "message": "Missing required field: "id"",
           "name": "JointStatusError",
           "status": 400,
         }
       `)
    })

    it('should return an empty collection when the requested resource does not exist', async () => {
      const spec = {
        modelName: 'Project',
        fields: [
          { name: 'id', type: 'Number', required: true, lookup: true },
          { name: 'name', type: 'String' },
          { name: 'brief_description', type: 'String' }
        ]
      }
      const input = {
        fields: {
          id: 999,
          name: 'Updated Name'
        }
      }

      const updated = await projectApp.updateMany(spec, input)
      expect(updated.models).toHaveLength(0)
    })

    it('should update the resource when the spec is satisfied', async () => {
      const spec = {
        modelName: 'Project',
        fields: [
          { name: 'is_internal', type: 'Boolean', required: true, lookup: true },
          { name: 'status_code', type: 'Number', required: true, lookup: true },
          { name: 'name', type: 'String' }
        ]
      }

      // Perform update
      const updated = await projectApp.updateMany(spec, {
        fields: { status_code: 5, is_internal: true, name: 'Updated Name' }
      })

      expect(updated.models).toHaveLength(2)

      // Due to a bug with property matchers in array the snapshot tested must be done in a loop
      // https://github.com/jestjs/jest/issues/9079
      updated.models.forEach((item) => {
        expect(item).toHaveProperty('attributes.name', 'Updated Name')
        expect(omitInternalFields(item.attributes)).toMatchSnapshot()
      })
    })

    it(`should support the "${ACTION.SPEC_FIELDS_OPT_OPERATORS}" option and update all resources matching the input`, async () => {
      const spec = {
        modelName: 'Project',
        fields: [
          { name: 'name', type: 'String', required: true, lookup: true, operators: ['contains'] },
          { name: 'status_code', type: 'Number' }
        ]
      }

      const input = {
        fields: {
          'name.contains': 'er',
          status_code: 3
        }
      }

      // Perform update
      const updated = await projectApp.updateMany(spec, input)

      expect(updated.models).toHaveLength(2)

      // Due to a bug with property matchers in array the snapshot tested must be done in a loop
      // https://github.com/jestjs/jest/issues/9079
      updated.models.forEach((item) => {
        expect(item).toHaveProperty('attributes.status_code', 3)
        expect(omitInternalFields(item.attributes)).toMatchSnapshot()
      })
    })

    it(`should support dynamic values on the "${ACTION.SPEC_FIELDS_OPT_DEFAULT_VALUE}" option (now, camelCase, kebabCase, snakeCase, pascalCase)`, async () => {
      vi.setSystemTime(new Date('2024-01-02T00:00:00.000Z'))
      const spec = {
        modelName: 'Project',
        fields: [
          { name: 'is_internal', type: 'Boolean', required: true, lookup: true },
          { name: 'alias', type: 'String', locked: true, defaultValue: '% camelCase(full_description) %' },
          { name: 'location', type: 'String', defaultValue: '% kebabCase(full_description) %' },
          { name: 'name', type: 'String', defaultValue: '% snakeCase(full_description) %' },
          { name: 'brief_description', type: 'String', defaultValue: '% pascalCase(full_description) %' },
          { name: 'started_at', type: 'String', defaultValue: '% now %' },
          { name: 'full_description', type: 'String' }
        ]
      }

      const updated = await projectApp.updateMany(spec, {
        fields: { is_internal: true, full_description: 'test This guy' }
      })
      expect(updated.models).toHaveLength(4)

      // Due to a bug with property matchers in array the snapshot tested must be done in a loop
      // https://github.com/jestjs/jest/issues/9079
      updated.models.forEach((item) => {
        expect(omitInternalFields(item.attributes)).toMatchSnapshot()
      })
    })

    it(`should support an "${ACTION.SPEC_AUTH_OWNER_CREDS}" authorization from a field on the looked-up item data`, async () => {
      const userContext = {
        is_logged_in: true,
        id: 4,
        external_id: '304',
        username: 'the_manic_edge',
        roles: [],
        profile_ids: [1, 2, 3]
      }
      const authContext = blogApp.prepareAuthContext(userContext)

      const spec = {
        modelName: 'Profile',
        fields: [
          { name: 'id', type: 'Number', required: true, lookup: true },
          { name: 'title', type: 'String' },
          { name: 'tagline', type: 'String' }
        ],
        auth: {
          rules: { owner: 'me' },
          ownerCreds: ['user_id => id']
        }
      }

      const input = {
        fields: {
          id: 1,
          title: 'A New Title for a New Day'
        },
        authContext
      }

      const data = await blogApp.updateMany(spec, input)
      expect(data.models).toHaveLength(1)
      expect(omitInternalFields(data.models[0].attributes)).toMatchInlineSnapshot(`
        {
          "avatar_url": null,
          "cid": "c14",
          "created_at": 2024-01-01T00:05:00.000Z,
          "description": null,
          "id": 1,
          "is_default": 1,
          "is_live": 1,
          "relations": {},
          "slug": "functional-fanatic",
          "tagline": "I don't have habits, I have algorithms.",
          "title": "A New Title for a New Day",
          "updated_at": 2024-01-01T00:05:00.000Z,
          "user_id": 4,
        }
      `)
    })

    it('should return in JSON API shape when payload format is set to "json-api"', async () => {
      const spec = {
        modelName: 'Project',
        fields: [
          { name: 'name', type: 'String', required: true, lookup: true, operators: ['contains'] },
          { name: 'status_code', type: 'Number' }
        ]
      }

      const input = {
        fields: {
          'name.contains': 'er',
          status_code: 3
        }
      }

      // Globally set...
      const globalLevel = await projectAppJsonApi.updateMany(spec, input)
      expect(globalLevel).toHaveProperty('data')
      expect(globalLevel.data).toHaveLength(2)

      // Due to a bug with property matchers in array the snapshot tested must be done in a loop
      // https://github.com/jestjs/jest/issues/9079
      globalLevel.data.forEach((item) => {
        expect(omitInternalFields(item.attributes)).toMatchSnapshot()
      })

      // Locally set...
      const methodLevel = await projectApp.updateMany(spec, input, 'json-api')
      expect(methodLevel).toHaveProperty('data')
      expect(methodLevel.data).toHaveLength(2)

      // Due to a bug with property matchers in array the snapshot tested must be done in a loop
      // https://github.com/jestjs/jest/issues/9079
      methodLevel.data.forEach((item) => {
        expect(omitInternalFields(item.attributes)).toMatchSnapshot()
      })
    })
  })
})
