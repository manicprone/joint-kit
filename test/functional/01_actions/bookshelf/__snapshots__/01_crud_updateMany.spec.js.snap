// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CRUD ACTIONS [bookshelf] > updateMany > should return in JSON API shape when payload format is set to "json-api" 1`] = `
{
  "alias": "mega-seed-mini-sythesizer",
  "brief_description": null,
  "cid": "c16",
  "created_at": 2024-01-01T00:05:00.000Z,
  "created_by": null,
  "finished_at": "2016-08-02T21:33+08:00",
  "full_description": null,
  "image_url": "http://vignette4.wikia.nocookie.net/rickandmorty/images/e/e8/Speed_Mega_Seed.png/revision/latest",
  "is_internal": 1,
  "location": null,
  "name": "Mega-Seed Mini-Sythesizer",
  "relations": {},
  "started_at": "2016-07-30T11:20+08:00",
  "status_code": 3,
  "updated_at": 2024-01-01T00:05:00.000Z,
}
`;

exports[`CRUD ACTIONS [bookshelf] > updateMany > should return in JSON API shape when payload format is set to "json-api" 2`] = `
{
  "alias": "doppelganger-finder",
  "brief_description": null,
  "cid": "c17",
  "created_at": 2024-01-01T00:15:00.000Z,
  "created_by": null,
  "finished_at": "2017-03-29",
  "full_description": null,
  "image_url": null,
  "is_internal": 1,
  "location": null,
  "name": "Doppelgänger Finder",
  "relations": {},
  "started_at": "2017-03-10",
  "status_code": 3,
  "updated_at": 2024-01-01T00:15:00.000Z,
}
`;

exports[`CRUD ACTIONS [bookshelf] > updateMany > should return in JSON API shape when payload format is set to "json-api" 3`] = `
{
  "alias": "mega-seed-mini-sythesizer",
  "brief_description": null,
  "cid": "c19",
  "created_at": 2024-01-01T00:05:00.000Z,
  "created_by": null,
  "finished_at": "2016-08-02T21:33+08:00",
  "full_description": null,
  "image_url": "http://vignette4.wikia.nocookie.net/rickandmorty/images/e/e8/Speed_Mega_Seed.png/revision/latest",
  "is_internal": 1,
  "location": null,
  "name": "Mega-Seed Mini-Sythesizer",
  "relations": {},
  "started_at": "2016-07-30T11:20+08:00",
  "status_code": 3,
  "updated_at": 2024-01-01T00:05:00.000Z,
}
`;

exports[`CRUD ACTIONS [bookshelf] > updateMany > should return in JSON API shape when payload format is set to "json-api" 4`] = `
{
  "alias": "doppelganger-finder",
  "brief_description": null,
  "cid": "c20",
  "created_at": 2024-01-01T00:15:00.000Z,
  "created_by": null,
  "finished_at": "2017-03-29",
  "full_description": null,
  "image_url": null,
  "is_internal": 1,
  "location": null,
  "name": "Doppelgänger Finder",
  "relations": {},
  "started_at": "2017-03-10",
  "status_code": 3,
  "updated_at": 2024-01-01T00:15:00.000Z,
}
`;

exports[`CRUD ACTIONS [bookshelf] > updateMany > should support dynamic values on the "defaultValue" option (now, camelCase, kebabCase, snakeCase, pascalCase) 1`] = `
{
  "alias": "testThisGuy",
  "brief_description": "TestThisGuy",
  "cid": "c9",
  "created_at": 2024-01-01T00:05:00.000Z,
  "created_by": null,
  "finished_at": "2016-08-02T21:33+08:00",
  "full_description": "test This guy",
  "id": 1,
  "image_url": "http://vignette4.wikia.nocookie.net/rickandmorty/images/e/e8/Speed_Mega_Seed.png/revision/latest",
  "is_internal": 1,
  "location": "test-this-guy",
  "name": "test_this_guy",
  "relations": {},
  "started_at": "2024-01-02T00:00:00Z",
  "status_code": 4,
  "updated_at": 2024-01-01T00:05:00.000Z,
}
`;

exports[`CRUD ACTIONS [bookshelf] > updateMany > should support dynamic values on the "defaultValue" option (now, camelCase, kebabCase, snakeCase, pascalCase) 2`] = `
{
  "alias": "testThisGuy",
  "brief_description": "TestThisGuy",
  "cid": "c10",
  "created_at": 2024-01-01T00:10:00.000Z,
  "created_by": null,
  "finished_at": null,
  "full_description": "test This guy",
  "id": 2,
  "image_url": "https://everydayaboverubies.files.wordpress.com/2011/11/jar_of_pickles_cb101311.jpg",
  "is_internal": 1,
  "location": "test-this-guy",
  "name": "test_this_guy",
  "relations": {},
  "started_at": "2024-01-02T00:00:00Z",
  "status_code": 5,
  "updated_at": 2024-01-01T00:10:00.000Z,
}
`;

exports[`CRUD ACTIONS [bookshelf] > updateMany > should support dynamic values on the "defaultValue" option (now, camelCase, kebabCase, snakeCase, pascalCase) 3`] = `
{
  "alias": "testThisGuy",
  "brief_description": "TestThisGuy",
  "cid": "c11",
  "created_at": 2024-01-01T00:15:00.000Z,
  "created_by": null,
  "finished_at": "2017-03-29",
  "full_description": "test This guy",
  "id": 3,
  "image_url": null,
  "is_internal": 1,
  "location": "test-this-guy",
  "name": "test_this_guy",
  "relations": {},
  "started_at": "2024-01-02T00:00:00Z",
  "status_code": 5,
  "updated_at": 2024-01-01T00:15:00.000Z,
}
`;

exports[`CRUD ACTIONS [bookshelf] > updateMany > should support dynamic values on the "defaultValue" option (now, camelCase, kebabCase, snakeCase, pascalCase) 4`] = `
{
  "alias": "testThisGuy",
  "brief_description": "TestThisGuy",
  "cid": "c12",
  "created_at": 2024-01-01T00:20:00.000Z,
  "created_by": null,
  "finished_at": null,
  "full_description": "test This guy",
  "id": 4,
  "image_url": "https://i.pinimg.com/736x/53/2e/e1/532ee1735e657073f4063a2cbed4e7f1--jello-popsicles-aqua-blue.jpg",
  "is_internal": 1,
  "location": "test-this-guy",
  "name": "test_this_guy",
  "relations": {},
  "started_at": "2024-01-02T00:00:00Z",
  "status_code": 3,
  "updated_at": 2024-01-01T00:20:00.000Z,
}
`;

exports[`CRUD ACTIONS [bookshelf] > updateMany > should support the "operators" option and update all resources matching the input 1`] = `
{
  "alias": "mega-seed-mini-sythesizer",
  "brief_description": null,
  "cid": "c6",
  "created_at": 2024-01-01T00:05:00.000Z,
  "created_by": null,
  "finished_at": "2016-08-02T21:33+08:00",
  "full_description": null,
  "id": 1,
  "image_url": "http://vignette4.wikia.nocookie.net/rickandmorty/images/e/e8/Speed_Mega_Seed.png/revision/latest",
  "is_internal": 1,
  "location": null,
  "name": "Mega-Seed Mini-Sythesizer",
  "relations": {},
  "started_at": "2016-07-30T11:20+08:00",
  "status_code": 3,
  "updated_at": 2024-01-01T00:05:00.000Z,
}
`;

exports[`CRUD ACTIONS [bookshelf] > updateMany > should support the "operators" option and update all resources matching the input 2`] = `
{
  "alias": "doppelganger-finder",
  "brief_description": null,
  "cid": "c7",
  "created_at": 2024-01-01T00:15:00.000Z,
  "created_by": null,
  "finished_at": "2017-03-29",
  "full_description": null,
  "id": 3,
  "image_url": null,
  "is_internal": 1,
  "location": null,
  "name": "Doppelgänger Finder",
  "relations": {},
  "started_at": "2017-03-10",
  "status_code": 3,
  "updated_at": 2024-01-01T00:15:00.000Z,
}
`;

exports[`CRUD ACTIONS [bookshelf] > updateMany > should update the resource when the spec is satisfied 1`] = `
{
  "alias": "turn-myself-into-a-pickle",
  "brief_description": "Turn myself into a pickle.",
  "cid": "c3",
  "created_at": 2024-01-01T00:10:00.000Z,
  "created_by": null,
  "finished_at": null,
  "full_description": null,
  "id": 2,
  "image_url": "https://everydayaboverubies.files.wordpress.com/2011/11/jar_of_pickles_cb101311.jpg",
  "is_internal": 1,
  "location": null,
  "name": "Updated Name",
  "relations": {},
  "started_at": null,
  "status_code": 5,
  "updated_at": 2024-01-01T00:10:00.000Z,
}
`;

exports[`CRUD ACTIONS [bookshelf] > updateMany > should update the resource when the spec is satisfied 2`] = `
{
  "alias": "doppelganger-finder",
  "brief_description": null,
  "cid": "c4",
  "created_at": 2024-01-01T00:15:00.000Z,
  "created_by": null,
  "finished_at": "2017-03-29",
  "full_description": null,
  "id": 3,
  "image_url": null,
  "is_internal": 1,
  "location": null,
  "name": "Updated Name",
  "relations": {},
  "started_at": "2017-03-10",
  "status_code": 5,
  "updated_at": 2024-01-01T00:15:00.000Z,
}
`;
